composer/models/bert/model.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Mapping, Tuple
+
+import transformers
+from torchmetrics import Accuracy, MatthewsCorrcoef, MeanSquaredError, SpearmanCorrcoef
+from torchmetrics.collections import MetricCollection
+
+from composer.models.nlp_metrics import BinaryF1Score, CrossEntropyLoss, MaskedAccuracy
+from composer.models.transformer_shared import MosaicTransformer
+
+if TYPE_CHECKING:
+    from composer.core.types import Batch, BatchDict, Metrics, Tensors
+
+
+class BERTModel(MosaicTransformer):
+    """
+    Implements a BERT wrapper around a MosaicTransformer.
+    """
+
+    def __init__(self, module: transformers.BertModel, config: transformers.BertConfig, tokenizer_name: str) -> None:
+        super().__init__(
+            module=module,  #type: ignore (thirdparty)
+            config=config,
+            tokenizer_name=tokenizer_name)
+
+        # we're going to remove the label from the expected inputs
+        # since we will handle metric calculation with TorchMetrics instead of HuggingFace.
+        self.model_inputs.remove("labels")
+
+        self.train_metrics = []
+        self.val_metrics = []
+
+        # TODO (Moin): make sure this is moved to be dataset-specific
+        # if config.num_labels=1, then we are training a regression task, so we should update our loss functions
+        if config.num_labels == 1:
+            self.train_loss = MeanSquaredError()
+            self.val_loss = MeanSquaredError()
+
+            self.train_spearman = SpearmanCorrcoef()
+            self.val_spearman = SpearmanCorrcoef()
+
+            self.train_metrics.extend([self.train_loss, self.train_spearman])
+            self.val_metrics.extend([self.val_loss, self.val_spearman])
+
+        if config.num_labels == 2:
+            self.train_f1 = BinaryF1Score()
+            self.val_f1 = BinaryF1Score()
+
+            self.train_metrics.extend([self.train_f1])
+            self.val_metrics.extend([self.val_f1])
+
+        if config.num_labels > 1 and config.num_labels != len(self.tokenizer):
+            self.train_acc = Accuracy()
+            self.val_acc = Accuracy()
+
+            self.train_matthews = MatthewsCorrcoef(num_classes=config.num_labels)
+            self.val_matthews = MatthewsCorrcoef(num_classes=config.num_labels)
+
+            self.train_metrics.extend([self.train_acc, self.train_matthews])
+            self.val_metrics.extend([self.val_acc, self.val_matthews])
+
+        if config.num_labels == len(self.tokenizer):  # tests for MLM pre-training
+            ignore_index = -100
+            self.train_loss = CrossEntropyLoss(ignore_index=ignore_index, vocab_size=config.num_labels)
```

Comment:
OK gotcha... so I guess there's some naming conventions to clean up because `training_metrics` is really like `batch_metrics` and the usual metrics are more like dataset `eval_metrics`... and then again batch metrics don't really need to be metrics at all... making my head spin but should be assigned in a separate issue :P